File: pos-scanner-web/backend/scan_service.py
import concurrent.futures
import ipaddress
import json
import logging
import socket
import requests
from typing import List, Dict, Any

# 配置日志
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class ScanPosService:
    def __init__(self, local_ip=None):
        self.local_ip = local_ip

    @staticmethod
    def guess_os_by_ip(ip, port=22080, timeout=3):
        url = f"http://{ip}:{port}/kpos/webapp/os/getOSType"
        try:
            response = requests.get(url, timeout=timeout)
            if response.status_code == 200:
                data = response.json()
                return data.get("os", "Unknown")
            return "Unknown"
        except Exception:
            return "Unknown"

    def fetch_company_profile(self, ip, port=22080, timeout=5, max_retries=2):
        url = f"http://{ip}:{port}/kpos/webapp/store/fetchCompanyProfile"
        last_error = None
        for attempt in range(max_retries):
            try:
                response = requests.get(url, timeout=timeout)
                if response.status_code == 200:
                    return response.json()
                last_error = f"HTTP {response.status_code}"
                logger.warning(f"获取设备信息失败 {ip}: {last_error} (尝试 {attempt + 1}/{max_retries})")
            except requests.exceptions.RequestException as e:
                last_error = f"Request error: {str(e)}"
                logger.warning(f"请求异常 {ip}: {last_error} (尝试 {attempt + 1}/{max_retries})")
            except json.JSONDecodeError as e:
                last_error = f"JSON decode error: {str(e)}"
                logger.warning(f"JSON解析失败 {ip}: {last_error} (尝试 {attempt + 1}/{max_retries})")
        logger.error(f"获取设备信息最终失败 {ip}: {last_error}")
        return {"error": f"Failed after {max_retries} retries: {last_error}"}

    def _scan_port(self, ip, port, timeout=2):
        """扫描指定IP的端口是否开放

        Args:
            ip: 目标IP地址
            port: 目标端口
            timeout: 连接超时时间（秒），默认2秒
        """
        try:
            with socket.create_connection((str(ip), port), timeout):
                return ip
        except socket.timeout:
            logger.debug(f"端口扫描超时: {ip}:{port}")
            return None
        except ConnectionRefusedError:
            logger.debug(f"连接被拒绝: {ip}:{port}")
            return None
        except Exception as e:
            logger.debug(f"端口扫描异常 {ip}:{port}: {type(e).__name__}")
            return None

    def _get_local_network(self):
        local_ip = self.local_ip if self.local_ip else socket.gethostbyname(socket.gethostname())
        return ipaddress.IPv4Network(f"{local_ip}/23", strict=False)

    def _extract_required_info(self, api_response):
        try:
            company = api_response.get("company", {})
            result = {
                "merchantId": company.get("merchantId"),
                "name": company.get("name"),
                "version": company.get("appInfo", {}).get("version"),
            }
            return {k: v for k, v in result.items() if v is not None} or {"error": "No required fields"}
        except Exception as e:
            return {"error": str(e)}

    def scan_network(self, port=22080):
        network = self._get_local_network()
        hosts = list(network.hosts())
        logger.info(f"开始扫描网络 {network}，共 {len(hosts)} 个IP")

        # 扫描开放端口
        open_ips = self._scan_open_ips(hosts, port)
        logger.info(f"端口扫描完成，发现 {len(open_ips)} 个开放端口")

        # 获取设备信息
        results = self._fetch_profiles(open_ips, port)
        success_count = sum(1 for r in results if r.get("status") == "success")
        logger.info(f"扫描完成，成功获取 {success_count}/{len(results)} 台设备信息")
        return results

    def _scan_open_ips(self, hosts, port):
        open_ips = []
        with concurrent.futures.ThreadPoolExecutor(max_workers=200) as executor:
            futures = {executor.submit(self._scan_port, ip, port): ip for ip in hosts}
            for future in concurrent.futures.as_completed(futures):
                if (result := future.result()):
                    open_ips.append(str(result))
                    logger.debug(f"发现开放端口: {result}:{port}")
        return open_ips

    def _fetch_profiles(self, open_ips, port):
        results = []
        with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:
            future_to_ip = {executor.submit(self._fetch_and_process, ip, port): ip for ip in open_ips}
            for future in concurrent.futures.as_completed(future_to_ip):
                result = future.result()
                if result:
                    results.append(result)
        return results

    def _fetch_and_process(self, ip, port):
        full_data = self.fetch_company_profile(ip, port)
        simple_data = self._extract_required_info(full_data)
        device_type = self.guess_os_by_ip(ip, port)

        return {
            "ip": ip,
            "merchantId": simple_data.get("merchantId", ""),
            "name": simple_data.get("name", ""),
            "version": simple_data.get("version", ""),
            "type": device_type,
            "status": "success" if "error" not in simple_data else "error",
            "error": simple_data.get("error", ""),
            "fullData": full_data
        }

    def get_local_ips(self):
        """获取本地所有IPv4地址"""
        ips = set()
        hostname = socket.gethostname()
        for info in socket.getaddrinfo(hostname, None):
            ip = info[4][0]
            if "." in ip and not ip.startswith("127."):
                ips.add(ip)
        return list(ips)

File: pos-scanner-web/backend-go/internal/services/scan_service.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// ScanStatus represents the current scan status
type ScanStatus struct {
	mu         sync.RWMutex
	IsScanning bool                      `json:"is_scanning"`
	Progress   int                       `json:"progress"`
	CurrentIP  string                    `json:"current_ip"`
	Results    []map[string]interface{}  `json:"results"`
	Error      string                    `json:"error"`
}

// ScanService handles network scanning operations
type ScanService struct {
	status *ScanStatus
	cancel context.CancelFunc
	mu     sync.Mutex
}

// NewScanService creates a new scan service
func NewScanService() *ScanService {
	return &ScanService{
		status: &ScanStatus{
			IsScanning: false,
			Progress:   0,
			Results:    make([]map[string]interface{}, 0),
		},
	}
}

// GetStatus returns the current scan status
func (s *ScanService) GetStatus() *ScanStatus {
	s.status.mu.RLock()
	defer s.status.mu.RUnlock()
	return &ScanStatus{
		IsScanning: s.status.IsScanning,
		Progress:   s.status.Progress,
		CurrentIP:  s.status.CurrentIP,
		Results:    s.status.Results,
		Error:      s.status.Error,
	}
}

// GetLocalIPs returns all local IPv4 addresses
func (s *ScanService) GetLocalIPs() ([]string, error) {
	ips := make([]string, 0)

	// Try hostname lookup first
	hostname, err := os.Hostname()
	if err == nil {
		hostnameIPs, err := net.LookupHost(hostname)
		if err == nil {
			for _, ip := range hostnameIPs {
				if net.ParseIP(ip).To4() != nil && !strings.HasPrefix(ip, "127.") {
					ips = append(ips, ip)
				}
			}
		}
	}

	// Fallback: get all network interfaces
	if len(ips) == 0 {
		ifaces, err := net.Interfaces()
		if err != nil {
			return nil, err
		}
		for _, iface := range ifaces {
			addrs, err := iface.Addrs()
			if err != nil {
				continue
			}
			for _, addr := range addrs {
				if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() && ipnet.IP.To4() != nil {
					ips = append(ips, ipnet.IP.String())
				}
			}
		}
	}

	// If still empty, try another approach
	if len(ips) == 0 {
		conn, err := net.Dial("udp", "8.8.8.8:80")
		if err == nil {
			localAddr := conn.LocalAddr().(*net.UDPAddr)
			ips = append(ips, localAddr.IP.String())
			conn.Close()
		}
	}

	return ips, nil
}

// StartScan initiates a network scan
func (s *ScanService) StartScan(localIP string, onResult func(result map[string]interface{})) error {
	s.mu.Lock()
	if s.status.IsScanning {
		s.mu.Unlock()
		return fmt.Errorf("scan already in progress")
	}

	// Reset status
	s.status.mu.Lock()
	s.status.IsScanning = true
	s.status.Progress = 0
	s.status.CurrentIP = ""
	s.status.Results = make([]map[string]interface{}, 0)
	s.status.Error = ""
	s.status.mu.Unlock()

	ctx, cancel := context.WithCancel(context.Background())
	s.cancel = cancel
	s.mu.Unlock()

	go s.performScan(ctx, localIP, onResult)

	return nil
}

// StopScan stops the current scan
func (s *ScanService) StopScan() {
	s.mu.Lock()
	if s.cancel != nil {
		s.cancel()
	}
	s.status.mu.Lock()
	s.status.IsScanning = false
	s.status.mu.Unlock()
	s.mu.Unlock()
}

// performScan executes the actual network scan
func (s *ScanService) performScan(ctx context.Context, localIP string, onResult func(result map[string]interface{})) {
	defer func() {
		s.status.mu.Lock()
		s.status.IsScanning = false
		s.status.mu.Unlock()
	}()

	// Get network range
	_, ipNet, err := net.ParseCIDR(localIP + "/23")
	if err != nil {
		s.status.mu.Lock()
		s.status.Error = err.Error()
		s.status.mu.Unlock()
		return
	}

	// Generate all hosts in the network
	hosts := generateHosts(ipNet)
	totalHosts := len(hosts)

	// Scan ports concurrently
	openIPs := make([]string, 0)
	var mu sync.Mutex
	var scannedCount int

	// Use worker pool pattern
	workerCount := 200
	ipChan := make(chan string, workerCount)
	resultChan := make(chan string, workerCount)

	var wg sync.WaitGroup

	// Start workers
	for i := 0; i < workerCount; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for ip := range ipChan {
				select {
				case <-ctx.Done():
					return
				default:
					if s.scanPort(ip, 22080, 2) {
						resultChan <- ip
					}
					// Update progress more smoothly
					mu.Lock()
					scannedCount++
					progress := scannedCount * 40 / totalHosts
					if progress > 40 {
						progress = 40
					}
					s.status.mu.Lock()
					s.status.Progress = progress
					s.status.CurrentIP = ip
					s.status.mu.Unlock()
					mu.Unlock()
				}
			}
		}()
	}

	// Collect results
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Send IPs to workers
	go func() {
		for _, ip := range hosts {
			select {
			case <-ctx.Done():
				break
			default:
				ipChan <- ip
			}
		}
		close(ipChan)
	}()

	// Collect open IPs with progress updates
	s.status.mu.Lock()
	s.status.Progress = 5
	s.status.mu.Unlock()

	for ip := range resultChan {
		mu.Lock()
		openIPs = append(openIPs, ip)
		scannedCount++
		progress := 5 + scannedCount*40/totalHosts
		if progress > 45 {
			progress = 45
		}
		s.status.mu.Lock()
		s.status.Progress = progress
		s.status.mu.Unlock()
		mu.Unlock()
	}

	// Port scanning complete
	s.status.mu.Lock()
	s.status.Progress = 45
	s.status.mu.Unlock()

	// Fetch device info concurrently
	totalOpen := len(openIPs)
	if totalOpen == 0 {
		totalOpen = 1
	}

	// Use worker pool for fetching device info
	fetchWorkerCount := 100 // 并发数
	if fetchWorkerCount > totalOpen {
		fetchWorkerCount = totalOpen
	}

	fetchChan := make(chan string, fetchWorkerCount)
	fetchResultChan := make(chan map[string]interface{}, fetchWorkerCount)
	var fetchWg sync.WaitGroup
	var completedCount int

	// Start fetch workers
	for i := 0; i < fetchWorkerCount; i++ {
		fetchWg.Add(1)
		go func() {
			defer fetchWg.Done()
			for ip := range fetchChan {
				select {
				case <-ctx.Done():
					return
				default:
					result := s.fetchAndProcess(ip, 22080)
					fetchResultChan <- result
				}
			}
		}()
	}

	// Collect results
	go func() {
		fetchWg.Wait()
		close(fetchResultChan)
	}()

	// Send IPs to fetch workers
	go func() {
		for _, ip := range openIPs {
			select {
			case <-ctx.Done():
				return
			default:
				fetchChan <- ip
			}
		}
		close(fetchChan)
	}()

	// Collect fetch results
	for result := range fetchResultChan {
		s.status.mu.Lock()
		s.status.Results = append(s.status.Results, result)
		completedCount++
		// Update progress (45-95%)
		s.status.Progress = 45 + completedCount*50/totalOpen
		if s.status.Progress > 95 {
			s.status.Progress = 95
		}
		if ip, ok := result["ip"].(string); ok {
			s.status.CurrentIP = ip
		}
		s.status.mu.Unlock()

		if onResult != nil {
			onResult(result)
		}
	}

	// Complete
	s.status.mu.Lock()
	s.status.Progress = 100
	s.status.CurrentIP = ""
	s.status.mu.Unlock()
}

// scanPort checks if a port is open on the given IP
func (s *ScanService) scanPort(ip string, port, timeout int) bool {
	address := net.JoinHostPort(ip, strconv.Itoa(port))
	conn, err := net.DialTimeout("tcp", address, time.Duration(timeout)*time.Second)
	if err != nil {
		return false
	}
	conn.Close()
	return true
}

// fetchAndProcess retrieves device information
func (s *ScanService) fetchAndProcess(ip string, port int) map[string]interface{} {
	fullData := s.fetchCompanyProfile(ip, port, 5, 2)
	simpleData := s.extractRequiredInfo(fullData)
	deviceType := s.guessOS(ip, port, 3)

	merchantID := ""
	if v, ok := simpleData["merchantId"].(string); ok {
		merchantID = v
	}

	status := "success"
	if _, hasError := simpleData["error"]; hasError {
		status = "error"
	}

	return map[string]interface{}{
		"ip":         ip,
		"merchantId": merchantID,
		"name":       simpleData["name"],
		"version":    simpleData["version"],
		"type":       deviceType,
		"status":     status,
		"error":      simpleData["error"],
		"fullData":   fullData,
	}
}

// fetchCompanyProfile retrieves the company profile from a device
func (s *ScanService) fetchCompanyProfile(ip string, port, timeout, maxRetries int) map[string]interface{} {
	url := fmt.Sprintf("http://%s:%d/kpos/webapp/store/fetchCompanyProfile", ip, port)

	for attempt := 0; attempt < maxRetries; attempt++ {
		client := &http.Client{Timeout: time.Duration(timeout) * time.Second}
		resp, err := client.Get(url)
		if err != nil {
			continue
		}
		defer resp.Body.Close()

		if resp.StatusCode == http.StatusOK {
			body, err := io.ReadAll(resp.Body)
			if err != nil {
				continue
			}

			var result map[string]interface{}
			if err := json.Unmarshal(body, &result); err != nil {
				continue
			}
			return result
		}
	}

	return map[string]interface{}{"error": "Failed after max retries"}
}

// guessOS determines the OS type of a device
func (s *ScanService) guessOS(ip string, port, timeout int) string {
	url := fmt.Sprintf("http://%s:%d/kpos/webapp/os/getOSType", ip, port)

	client := &http.Client{Timeout: time.Duration(timeout) * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return "Unknown"
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return "Unknown"
		}

		var result map[string]interface{}
		if err := json.Unmarshal(body, &result); err != nil {
			return "Unknown"
		}

		if os, ok := result["os"].(string); ok {
			return os
		}
	}

	return "Unknown"
}

// extractRequiredInfo extracts required fields from API response
func (s *ScanService) extractRequiredInfo(apiResponse map[string]interface{}) map[string]interface{} {
	company, ok := apiResponse["company"].(map[string]interface{})
	if !ok {
		return map[string]interface{}{"error": "No company field"}
	}

	result := make(map[string]interface{})
	if merchantID, ok := company["merchantId"].(string); ok {
		result["merchantId"] = merchantID
	}
	if name, ok := company["name"].(string); ok {
		result["name"] = name
	}
	if appInfo, ok := company["appInfo"].(map[string]interface{}); ok {
		if version, ok := appInfo["version"].(string); ok {
			result["version"] = version
		}
	}

	if len(result) == 0 {
		return map[string]interface{}{"error": "No required fields"}
	}

	return result
}

// generateHosts generates all host IPs in a network
func generateHosts(ipNet *net.IPNet) []string {
	hosts := make([]string, 0)

	ip := make(net.IP, len(ipNet.IP))
	copy(ip, ipNet.IP)

	for ip := ip.Mask(ipNet.Mask); ipNet.Contains(ip); inc(ip) {
		// Skip network and broadcast addresses
		if !ip.Equal(ipNet.IP.Mask(ipNet.Mask)) && !ip.Equal(broadcast(ipNet)) {
			hosts = append(hosts, ip.String())
		}
	}

	return hosts
}

// inc increments an IP address
func inc(ip net.IP) {
	for j := len(ip) - 1; j >= 0; j-- {
		ip[j]++
		if ip[j] > 0 {
			break
		}
	}
}

// broadcast returns the broadcast address of a network
func broadcast(ipNet *net.IPNet) net.IP {
	ip := make(net.IP, len(ipNet.IP))
	copy(ip, ipNet.IP)

	for i := 0; i < len(ip); i++ {
		ip[i] |= ^ipNet.Mask[i]
	}

	return ip
}

// FetchDeviceDetails fetches detailed information for a specific device
func (s *ScanService) FetchDeviceDetails(ip string) (map[string]interface{}, error) {
	return s.fetchCompanyProfile(ip, 22080, 5, 2), nil
}
